libclc/ptx-nvidiacl/libspirv/synchronization/barrier.cl

```diff
@@ -9,14 +9,33 @@
 #include <spirv/spirv.h>
 #include <spirv/spirv_types.h>
 
+int __clc_nvvm_reflect_arch();
+
 _CLC_OVERLOAD _CLC_DEF void __spirv_MemoryBarrier(unsigned int memory,
                                                   unsigned int semantics) {
-  if (memory == CrossDevice) {
-    __nvvm_membar_sys();
-  } else if (memory == Device) {
-    __nvvm_membar_gl();
-  } else {
-    __nvvm_membar_cta();
+
+  // for sm_70 and above membar becomes semantically identical to fence.sc.
+  // However sm_70 and above also introduces a lightweight fence.acq_rel that
+  // can be used to form either acquire or release strong operations.
+
+  unsigned int order = semantics & 0x1F;
+  if (__clc_nvvm_reflect_arch() < 700 ||
+             order == SequentiallyConsistent) {
+    if (memory == CrossDevice) {
+      __nvvm_membar_sys();
+    } else if (memory == Device) {
+      __nvvm_membar_gl();
+    } else {
+      __nvvm_membar_cta();
+    }
+  } else if (order != None) {
+    if (memory == CrossDevice) {
+      __asm__ __volatile__("fence.acq_rel.sys;");
```

Comment:
I am assuming that there are no clang builtins for these instructions?

Reply:
Yeah that's right. I was going to add a clang builtin that takes a semantic as an argument and then picks the correct instruction based on such an argument, but there are already some of these instructions implemented as builtins with a 1:1 relationship of builtin and complete instruction. Since I think this instruction is only ever going to be called by us in this particular code, I didn't see much point of adding builtins.

Reply:
Couldn't find any after a bit of searching. I am a bit surprised, might be worth mentioning upstream?

Reply:
Yeah fair enough

Reply:
Saying all this if anyone thinks I should add builtins for it I'm happy to do it.

My beginners understanding on this is that builtins are useful for
- when compiler can do special stuff with them based on the builtin, too much inline asm prevents optimizations
- simplifying instructions to deal with multiple architecture potentially (amd for example)
- simplifying instructions by mapping many to a single builtin that takes args specifying the instruction
- easier to type

I think upstream sometimes they don't add builtins for certain things if there isn't much value to it, one example is activemask which they just implement in asm.

Reply:
yeah OK. Thanks